src/sources/aws_ecs_metrics/mod.rs

```diff
@@ -0,0 +1,584 @@
+use crate::{
+    config::{self, GenerateConfig, GlobalOptions, SourceConfig, SourceDescription},
+    internal_events::{
+        AwsEcsMetricsErrorResponse, AwsEcsMetricsHttpError, AwsEcsMetricsParseError,
+        AwsEcsMetricsReceived, AwsEcsMetricsRequestCompleted,
+    },
+    shutdown::ShutdownSignal,
+    Event, Pipeline,
+};
+use futures::{compat::Sink01CompatExt, future, stream, FutureExt, StreamExt, TryFutureExt};
+use futures01::Sink;
+use hyper::{Body, Client, Request};
+use serde::{Deserialize, Serialize};
+use std::{
+    env,
+    time::{Duration, Instant},
+};
+
+pub mod parser;
+
+#[derive(Deserialize, Serialize, Clone, Debug)]
+#[serde(rename_all = "lowercase")]
+enum Version {
+    V2,
+    V3,
+    V4,
+}
+
+#[derive(Deserialize, Serialize, Clone, Debug)]
+struct AwsEcsMetricsSourceConfig {
+    version: Version,
+    #[serde(default = "default_scrape_interval_secs")]
+    scrape_interval_secs: u64,
+}
+
+pub fn default_scrape_interval_secs() -> u64 {
+    15
+}
+
+inventory::submit! {
+    SourceDescription::new::<AwsEcsMetricsSourceConfig>("aws_ecs_metrics")
+}
+
+impl GenerateConfig for AwsEcsMetricsSourceConfig {
+    fn generate_config() -> toml::Value {
+        toml::Value::try_from(Self {
+            version: Version::V4,
+            scrape_interval_secs: default_scrape_interval_secs(),
+        })
+        .unwrap()
+    }
+}
+
+#[async_trait::async_trait]
+#[typetag::serde(name = "aws_ecs_metrics")]
+impl SourceConfig for AwsEcsMetricsSourceConfig {
+    async fn build(
+        &self,
+        _name: &str,
+        _globals: &GlobalOptions,
+        shutdown: ShutdownSignal,
+        out: Pipeline,
+    ) -> crate::Result<super::Source> {
+        let url = match self.version {
+            Version::V2 => "169.254.170.2/v2/stats".to_string(),
+            Version::V3 => format!("{}/task/stats", env::var("ECS_CONTAINER_METADATA_URI")?),
+            Version::V4 => format!("{}/task/stats", env::var("ECS_CONTAINER_METADATA_URI_V4")?),
+        };
+
+        Ok(aws_ecs_metrics(
+            url,
+            self.scrape_interval_secs,
+            shutdown,
+            out,
+        ))
+    }
+
+    fn output_type(&self) -> crate::config::DataType {
+        config::DataType::Metric
+    }
+
+    fn source_type(&self) -> &'static str {
+        "aws_ecs_metrics"
+    }
+}
+
+fn aws_ecs_metrics(
+    url: String,
+    interval: u64,
+    shutdown: ShutdownSignal,
+    out: Pipeline,
+) -> super::Source {
+    let out = out
+        .sink_map_err(|e| error!("error sending metric: {:?}", e))
+        .sink_compat();
+    let task = tokio::time::interval(Duration::from_secs(interval))
+        .take_until(shutdown)
+        .map(move |_| {
+            let client = Client::new();
+
+            let request = Request::get(&url)
+                .body(Body::empty())
+                .expect("error creating request");
+
+            let start = Instant::now();
+            let url2 = url.clone();
```

Comment:
I'm taking this was necessary, but I'm not seeing why.

Reply:
make it a ref and avoided cloning.